### ISSUE TITLE
How do prototypes learn？

### ISSUE THREAD

**Issue by BAOOOOOM** [szzexpoi/POEM#1]
Thank you very much for your work, it was very interesting! But I'm curious about how prototypes learn. 
The article says that prototypes can be learned, represented as a linear layer in the code. However, since the initial state is equivalent to random, should we let it learn gradually? Why can we get object representations like Table 4 by learning in this way?

**Comment by szzexpoi** [comment]
Thanks for your interest in our study!

As described in the paper, our method consists of two stages of training, i.e., learning the prototypes and leveraging the prototypes (fixed) for visual reasoning. In the first stage, the prototypes (represented as weights in the linear layers) are first randomly initialized and then trained on a multi-label object classification task, where the predictions are computed based on an adaptive composition of prototypes. In this way, we can learn prototypes representing various objects such as those visualized in Figure 4. For more details, please refer to the "proto_learning" folder.

**Comment by kikiclare** [comment]
I am very interested in your research. While going through the code, I encountered some issues. Firstly, in the prototype decomposition section, is there no design for NLP prototype decomposition? Another question is in proto_learning/dataloader.py: self.obj2idx = json.load(open(os.path.join(data_dir, 'obj2idx_gqa.json'))). I couldn't find the code that generates the 'obj2idx_gqa.json' file.

**Comment by szzexpoi** [comment]
> I am very interested in your research. While going through the code, I encountered some issues. Firstly, in the prototype decomposition section, is there no design for NLP prototype decomposition? Another question is in proto_learning/dataloader.py: self.obj2idx = json.load(open(os.path.join(data_dir, 'obj2idx_gqa.json'))). I couldn't find the code that generates the 'obj2idx_gqa.json' file.

You are right that there is no prototype on language modality, as the study focuses on learning compositional visual representation. The "obj2idx_gqa" file can be downloaded from the preprocessed annotation link. These annotations are essentially derived from the GQA raw annotations to formulate the multi-label object classification task (i.e., simultaneously predicting all objects within an image), for instance, "obj2idx_gqa" is a dictionary mapping the object labels to indices.

**Comment by bonbonx01** [comment]
The processed annotations you provided, I'm confused about its context.

1. How do you obtain **objects** in the obj2idx.json?  
2.  And in _prototype_data.json_ the data struction looks  like '2316444': {'ocean': [3], 'wetsuit': [5], 'hair': [0]};The '2316444' stands for the img_id,however, i'm not sure about what the {'ocean': [3], 'wetsuit': [5], 'hair': [0]} stands for and how to generator this ?Can you provide more information or code for generating the annotations?

**Comment by szzexpoi** [comment]
(1) The objects in obj2idx are either defined in the scene graphs of GQA dataset or as the 80 COCO objects in VQA dataset.

(2) The JSON stores a mapping between each image and the objects within it. You can reproduce the mapping by traversing the GQA scene graphs or the object detection annotations for MSCOCO. Regarding the numbers, in some preliminary experiments, I learned the prototypes with region-wise annotations and the numbers indicate the location of objects within the 36 regions in the bottom-up features. The results are suboptimal and thus I switched to instance-wise annotations with attention. It is hard for me to find that code now as I graduated a year ago.

**Comment by bonbonx01** [comment]
Sorry for taking so long to reply.
i am not sure what you mean by "instance-wise annotations with attention". What does it mean?

**Comment by szzexpoi** [comment]
Well, you can take a look at the paper for details. Essentially, you use spatial attention to aggregate prediction from different regions, so that there is no requirement for object annotations of each region.
